cpp/src/arrow/type.h

```diff
@@ -1886,6 +1902,15 @@ class ARROW_EXPORT FieldRef : public util::EqualityComparable<FieldRef> {
     }
     return out;
   }
+  template <typename T>
+  std::vector<GetType<T>> GetAllFlattened(const T& root,
+                                          MemoryPool* pool = NULLPTR) const {
+    std::vector<GetType<T>> out;
+    for (const auto& match : FindAll(root)) {
+      out.push_back(match.GetFlattened(root, pool).ValueOrDie());
```

Comment:
Uh. Unlike `GetAll`, this can fail for various reasons such as failure to allocate enough memory. In that case, we'd probably want to return an error instead of dying out.

Reply:
In that case, I'm a little confused as to why the non-flattened variants don't forward those errors either - since the standard `FieldPath::Get` methods can also fail (which was true prior to this PR).

I'm mostly referring to `GetOne` and `GetOneAndNone` here, as they already return a `Result`. Regardless, I'll propagate those errors for the new methods.